File: demo/midgard-sdk/src/payout.ts
import { Data } from "@lucid-evolution/lucid";
import { AddressSchema, OutputReferenceSchema, ValueSchema } from "@/common.js";
import { CardanoDatumSchema } from "@/ledger-state.js";

export const PayoutDatumSchema = Data.Object({
  l2_value: ValueSchema,
  l1_address: AddressSchema,
  l1_datum: CardanoDatumSchema,
});
export type PayoutDatum = Data.Static<typeof PayoutDatumSchema>;
export const PayoutDatum = PayoutDatumSchema as unknown as PayoutDatum;

export const PayoutSpendRedeemerSchema = Data.Enum([
  Data.Object({
    AddFunds: Data.Object({
      payout_input_index: Data.Integer(),
      payout_output_index: Data.Integer(),
      reserve_input_index: Data.Integer(),
      reserve_change_output_index: Data.Nullable(Data.Integer()),
      reserve_spend_redeemer_index: Data.Integer(),
      payout_spend_redeemer_index: Data.Integer(),
      hub_ref_input_index: Data.Integer(),
    }),
  }),
  Data.Object({
    ConcludeWithdrawal: Data.Object({
      payout_input_index: Data.Integer(),
      l1_output_index: Data.Integer(),
      burn_redeemer_index: Data.Integer(),
      hub_ref_input_index: Data.Integer(),
    }),
  }),
]);
export type PayoutSpendRedeemer = Data.Static<typeof PayoutSpendRedeemerSchema>;
export const PayoutSpendRedeemer =
  PayoutSpendRedeemerSchema as unknown as PayoutSpendRedeemer;

export const PayoutMintRedeemerSchema = Data.Enum([
  Data.Object({
    MintPayout: Data.Object({
      withdrawal_utxo_out_ref: OutputReferenceSchema,
      withdrawal_input_index: Data.Integer(),
      withdrawal_spend_redeemer_index: Data.Integer(),
      hub_ref_input_index: Data.Integer(),
    }),
  }),
  Data.Object({
    BurnPayout: Data.Object({
      payout_input_index: Data.Integer(),
      payout_asset_name: Data.Bytes(),
      payout_spend_redeemer_index: Data.Integer(),
      hub_ref_input_index: Data.Integer(),
    }),
  }),
]);
export type PayoutMintRedeemer = Data.Static<typeof PayoutMintRedeemerSchema>;
export const PayoutMintRedeemer =
  PayoutMintRedeemerSchema as unknown as PayoutMintRedeemer;

File: onchain/plutarch/src/Midgard/Payout.hs
{- |
Module      : Midgard.Payout
Description : Partial Plutarch port of @lib/midgard/payout.ak@.

Ported so far: the datum and both redeemers. The reserve validator reads the
spend redeemer to check the two agree on which inputs they are pairing, and the
withdrawal validator reads the mint redeemer when initialising an accumulator.
The payout validator itself (1,436 lines) is a separate slice.
-}
module Midgard.Payout (
  PSpendRedeemer (..),
  PMintRedeemer (..),
  PPayoutDatum (..),
) where

import GHC.Generics (Generic)
import Generics.SOP qualified as SOP
import Plutarch.LedgerApi.Utils (PMaybeData)
import Plutarch.LedgerApi.V3 (PAddress, POutputDatum, PTokenName, PTxOutRef)
import Plutarch.Prelude

import Midgard.Common.Types (PValuePairs)

{- | Aiken @payout.SpendRedeemer@.

Constructor order fixes the on-chain tag: @AddFunds@ is 0, @ConcludeWithdrawal@
is 1. The reserve validator accepts only @AddFunds@, so a swapped order would
silently let a withdrawal drain the reserve.
-}
data PSpendRedeemer (s :: S)
  = PAddFunds
      { ppayoutAddFunds'payoutInputIndex :: Term s (PAsData PInteger)
      , ppayoutAddFunds'payoutOutputIndex :: Term s (PAsData PInteger)
      , ppayoutAddFunds'reserveInputIndex :: Term s (PAsData PInteger)
      , ppayoutAddFunds'reserveChangeOutputIndex :: Term s (PMaybeData PInteger)
      , ppayoutAddFunds'reserveSpendRedeemerIndex :: Term s (PAsData PInteger)
      , ppayoutAddFunds'payoutSpendRedeemerIndex :: Term s (PAsData PInteger)
      , ppayoutAddFunds'hubRefInputIndex :: Term s (PAsData PInteger)
      }
  | PConcludeWithdrawal
      { ppayoutConclude'payoutInputIndex :: Term s (PAsData PInteger)
      , ppayoutConclude'l1OutputIndex :: Term s (PAsData PInteger)
      , ppayoutConclude'burnRedeemerIndex :: Term s (PAsData PInteger)
      , ppayoutConclude'hubRefInputIndex :: Term s (PAsData PInteger)
      }
  deriving stock (Generic)
  deriving anyclass (SOP.Generic, PIsData, PEq, PShow)
  deriving (PlutusType) via (DeriveAsDataStruct PSpendRedeemer)

{- | Aiken @payout.MintRedeemer@.

Tags: @MintPayout@ 0, @BurnPayout@ 1.

@MintPayout@ is read by the withdrawal validator, which cross-checks every field
against its own redeemer — each script is relying on the other's indices, so
neither may be taken on trust.
-}
data PMintRedeemer (s :: S)
  = PMintPayout
      { pmintPayout'withdrawalUtxoOutRef :: Term s (PAsData PTxOutRef)
      , pmintPayout'withdrawalInputIndex :: Term s (PAsData PInteger)
      , pmintPayout'withdrawalSpendRedeemerIndex :: Term s (PAsData PInteger)
      , pmintPayout'hubRefInputIndex :: Term s (PAsData PInteger)
      }
  | PBurnPayout
      { pburnPayout'payoutInputIndex :: Term s (PAsData PInteger)
      , pburnPayout'payoutAssetName :: Term s (PAsData PTokenName)
      , pburnPayout'payoutSpendRedeemerIndex :: Term s (PAsData PInteger)
      , pburnPayout'hubRefInputIndex :: Term s (PAsData PInteger)
      }
  deriving stock (Generic)
  deriving anyclass (SOP.Generic, PIsData, PEq, PShow)
  deriving (PlutusType) via (DeriveAsDataStruct PMintRedeemer)

{- | Aiken @payout.Datum@.

What the accumulator is working towards: the L2 value owed, and where on L1 it
must eventually land.
-}
data PPayoutDatum (s :: S) = PPayoutDatum
  { ppayoutDatum'l2Value :: Term s (PAsData PValuePairs)
  , ppayoutDatum'l1Address :: Term s (PAsData PAddress)
  , ppayoutDatum'l1Datum :: Term s POutputDatum
  }
  deriving stock (Generic)
  deriving anyclass (SOP.Generic, PIsData, PEq, PShow)
  deriving (PlutusType) via (DeriveAsDataStruct PPayoutDatum)
